# Wrangle and Analyze Data
## WeRateDogs (@dog_rates)
### Introduction to the dataset
The dataset is the archive of tweets of the Twitter account @dog_rates, also known as WeRateDogs. WeRateDogs rates people's dogs with a humorous comment about the dog. Fun fact? The ratings almost always have a denominator of 10. However, the numerators are almost always greater than 10. The reason? *Because "they're good dogs Brent."*
WeRateDogs has over 4 million followers and has received international media coverage.
### Software
1. Jupyter Notebook
2. Packages:
    * pandas
    * NumPy
    * requests
    * json
    * tweepy

### Project details
Tasks:
1. Data wrangling, which consists of:
    * Gathering data (downloadable file in the Resources tab in the left most panel of your classroom and linked in step 1 below).
    * Assessing data
    * Cleaning data
2. Storing, analyzing, and visualizing your wrangled data
3. Reporting on 1) your data wrangling efforts and 2) your data analyses and visualizations


### Key Points
Key points to keep in mind when data wrangling for this project:

* You only want original ratings (no retweets) that have images. Though there are 5000+ tweets in the dataset, not all are dog ratings and some are retweets.
* Assessing and cleaning the entire dataset completely would require a lot of time, and is not necessary to practice and demonstrate your skills in data wrangling. Therefore, the requirements of this project are only to assess and clean at least 8 quality issues and at least 2 tidiness issues in this dataset.
* Cleaning includes merging individual pieces of data according to the rules of tidy data.
* The fact that the rating numerators are greater than the denominators does not need to be cleaned. This unique rating system is a big part of the popularity of WeRateDogs.
* You do not need to gather the tweets beyond August 1st, 2017. You can, but note that you won't be able to gather the image predictions for these tweets since you don't have access to the algorithm used.

## Gather

Data files provided by Udacity:

1. The WeRateDogs Twitter archive: 'twitter_archive_enhanced.csv'

2. The tweet image predictions: 'image-predictions.tsv'

3. Twitter API and Python's Tweepy library containing information about tweets: 'tweet-json'

In [1]:
# Packages import

import pandas as pd
import numpy as np
import requests
import json
import random

import matplotlib.pyplot as plt
from matplotlib import cm
%matplotlib inline

import seaborn as sns
sns.set(style="darkgrid")

# I don't have the Twitter API access, so I will skip this part and use 'tweet-json' file provide

In [2]:
# Read WeRateDogs Twitter archive

dogs_archive = pd.read_csv('twitter-archive-enhanced.csv')
dogs_archive.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [3]:
# Request downloading the image prediction using URL provided by Udacity

url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
image_request = requests.get(url, allow_redirects=True)

open('image-predictions.tsv', 'wb').write(image_request.content)

335079

In [4]:
# Reading the image predictions file

dogs_images = pd.read_csv('image-predictions.tsv', sep = '\t')
dogs_images.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [5]:
# Opening the text file provided by Udacity to get the 'tweet-jon' data

with open('tweet-json', mode='r') as file: print(file.readline())

{"created_at": "Tue Aug 01 16:23:56 +0000 2017", "id": 892420643555336193, "id_str": "892420643555336193", "full_text": "This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU", "truncated": false, "display_text_range": [0, 85], "entities": {"hashtags": [], "symbols": [], "user_mentions": [], "urls": [], "media": [{"id": 892420639486877696, "id_str": "892420639486877696", "indices": [86, 109], "media_url": "http://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg", "media_url_https": "https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg", "url": "https://t.co/MgUWQ76dJU", "display_url": "pic.twitter.com/MgUWQ76dJU", "expanded_url": "https://twitter.com/dog_rates/status/892420643555336193/photo/1", "type": "photo", "sizes": {"large": {"w": 540, "h": 528, "resize": "fit"}, "thumb": {"w": 150, "h": 150, "resize": "crop"}, "small": {"w": 540, "h": 528, "resize": "fit"}, "medium": {"w": 540, "h": 528, "resize": "fit"}}}]}, "extended_entities": {"media

In [6]:
json_data = [json.loads(line) for line in open('tweet-json', 'r')]

In [7]:
twitter_counts_df = pd.DataFrame(json_data)
twitter_counts_df = twitter_counts_df[['id', 'favorite_count', 'retweet_count']]
twitter_counts_df

id  favorite_count  retweet_count
0     892420643555336193           39467           8853
1     892177421306343426           33819           6514
2     891815181378084864           25461           4328
3     891689557279858688           42908           8964
4     891327558926688256           41048           9774
...                  ...             ...            ...
2349  666049248165822465             111             41
2350  666044226329800704             311            147
2351  666033412701032449             128             47
2352  666029285002620928             132             48
2353  666020888022790149            2535            532

[2354 rows x 3 columns]

In [8]:
twitter_counts_df.to_csv('tweet_data_extract.csv', index = False)

# Assess
In this section, we will proceed to assess the gathered data for quality and tidiness issues, bearing in mind the key points provided by Udacity -only tweets with original ratings with images, no retweets, be retained-.

In [9]:
# Load the data gathered data files
archive = pd.read_csv('twitter-archive-enhanced.csv')
predictions = pd.read_csv('image-predictions.tsv', sep = '\\t', engine = 'python')
json_data = pd.read_csv('tweet_data_extract.csv')

## Archive

In [10]:
archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [11]:
print("Data type for timestamp:  " + str(type(archive.timestamp[0])))

Data type for timestamp:  <class 'str'>


In [12]:
print("Duplicated:  {}".format(sum(archive.duplicated())))

Duplicated:  0


In [13]:
archive.head(10)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   
5  891087950875897856                    NaN                  NaN   
6  890971913173991426                    NaN                  NaN   
7  890729181411237888                    NaN                  NaN   
8  890609185150312448                    NaN                  NaN   
9  890240255349198849                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   
5  2017-07-29 00:08:17 +0000   
6  2017-07-28 16:27:12 +0000   
7  2017-07-28 00:22:40 +0000   
8  2017-07-27 16:25:51 +0000   
9  2017-07-26 15:59:51 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   
5  <a href="http://twitter.com/download/iphone" r...   
6  <a href="http://twitter.com/download/iphone" r...   
7  <a href="http://twitter.com/download/iphone" r...   
8  <a href="http://twitter.com/download/iphone" r...   
9  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   
5  Here we have a majestic great white breaching ...                  NaN   
6  Meet Jax. He enjoys ice cream so much he gets ...                  NaN   
7  When you watch your owner call another dog a g...                  NaN   
8  This is Zoey. She doesn't want to be one of th...                  NaN   
9  This is Cassie. She is a college pup. Studying...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   
5                       NaN                        NaN   
6                       NaN                        NaN   
7                       NaN                        NaN   
8                       NaN                        NaN   
9                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   
5  https://twitter.com/dog_rates/status/891087950...                13   
6  https://gofundme.com/ydvmve-surgery-for-jax,ht...                13   
7  https://twitter.com/dog_rates/status/890729181...                13   
8  https://twitter.com/dog_rates/status/890609185...                13   
9  https://twitter.com/dog_rates/status/890240255...    

In [14]:
archive.tail(10)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN                  NaN   
2350  666050758794694657                    NaN                  NaN   
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
2346  2015-11-16 01:01:59 +0000   
2347  2015-11-16 00:55:59 +0000   
2348  2015-11-16 00:49:46 +0000   
2349  2015-11-16 00:35:11 +0000   
2350  2015-11-16 00:30:50 +0000   
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
2346  <a href="http://twitter.com/download/iphone" r...   
2347  <a href="http://twitter.com/download/iphone" r...   
2348  <a href="http://twitter.com/download/iphone" r...   
2349  <a href="http://twitter.com/download/iphone" r...   
2350  <a href="http://twitter.com/download/iphone" r...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
2346  Here is the Rand Paul of retrievers folks! He'...                  NaN   
2347  My oh my. This is a rare blond Canadian terrie...                  NaN   
2348  Here is a Siberian heavily armored polar bear ...                  NaN   
2349  This is an odd dog. Hard on the outside but lo...                  NaN   
2350  This is a truly beautiful English Wilson Staff...                  NaN   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
2346                       NaN                        NaN   
2347                       NaN                        NaN   
2348                       NaN                        NaN   
2349                       NaN                        NaN   
2350                       NaN                        NaN   
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
2346  https://twitter.com/dog_rates/status/666058600...                 8   
2347  https://twitter.com/dog_rates/status/666057090...                 9   
2348  https://twitter.com/dog_rates/status/666055525...                10   
2349  https://twitter.com/dog_rates/status/666051853...                 2   
2350  https://twitter.com/dog_rates/status/666050758...                10   
2351  https://twitter.com/dog_rates/status/666049248...                 5   
2352  https://twitter.com/dog_rates/status/666044226...                 6   
2353  https://twit

#### Notes
The info() method reveals several quality and tidiness issues:

* There are 181 retweets (retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp)
* There are 78 replies (in_reply_to_status_id, in_reply_to_user_id)
* There are 2297 tweets with expanded_urls (links to images) hence, 59 tweets with missing data
* The timestamp field had string format (object)
* There are 4 columns for dog stages (doggo, floofer, pupper, puppo) and wrong category for stages
* The columns related to retweets are not applicable for original tweets
* The columns related to replies are not applicable for original tweets

The head() and tail() methods show us several issues:

Quality:

* The timestamp column has dates in string form.
* Some of the rows from the tail() output above have invalid strings in the name column, e.g. "a", "an", "in". These words are * all the 3rd word in the tweet.
* Values of "None" in the name column.

Tidiness:

* The columns with numerical data that are typically used for analysis are located to the far right of the table, and the columns with long strings are on the left; this makes it difficult to readily see the data that will be used for analyses.


#### Columns assessment

In [15]:
# Cheching 'name'

archive.name.value_counts().head(50)

None       745
a           55
Charlie     12
Cooper      11
Oliver      11
Lucy        11
Tucker      10
Lola        10
Penny       10
Winston      9
Bo           9
Sadie        8
the          8
Toby         7
an           7
Daisy        7
Bailey       7
Buddy        7
Bella        6
Koda         6
Rusty        6
Stanley      6
Jack         6
Milo         6
Oscar        6
Scout        6
Leo          6
Jax          6
Dave         6
very         5
Larry        5
Phil         5
George       5
Bentley      5
Louis        5
Sunny        5
Alfie        5
Sammy        5
Gus          5
Chester      5
Oakley       5
Finn         5
Boomer       4
Jeffrey      4
Beau         4
Walter       4
quite        4
Brody        4
Moose        4
Jerry        4
Name: name, dtype: int64

We can assume that the dog names are all capitalized, so words that begin in lowercase are probably not names, like "a", "the" and "an". Here's the list of these "names".

In [16]:
# Finding non-capitalised values
nc_names_mask = archive.name.str.contains('^[a-z]', regex = True)
archive[nc_names_mask].name.value_counts().sort_index()

a               55
actually         2
all              1
an               7
by               1
getting          2
his              1
incredibly       1
infuriating      1
just             4
life             1
light            1
mad              2
my               1
not              2
officially       1
old              1
one              4
quite            4
space            1
such             1
the              8
this             1
unacceptable     1
very             5
Name: name, dtype: int64

In [17]:
len(archive[nc_names_mask])

109

This list contains non-capitalized words from the name column for 109 tweets, none of which can be considered as a valid name.

In [18]:
# Checking 'expanded_urls' number of tweets with missing urls

sum(archive.expanded_urls.isna())

59

In [19]:
# Counting of replies and retweets with missing url

sum(archive.expanded_urls.isna() & \
    (archive.in_reply_to_status_id.notnull() | \
     archive.retweeted_status_id.notnull()))

56

In [20]:
# Checking tweets with missing url, filtering out replies and retweets

mask = archive.expanded_urls.isna() & \
       archive.in_reply_to_status_id.isna() & \
       archive.retweeted_status_id.isna()
archive[mask][['tweet_id', 'text', 'expanded_urls']]

tweet_id                                               text  \
375   828361771580813312  Beebop and Doobert should start a band 12/10 w...   
707   785515384317313025     Today, 10/10, should be National Dog Rates Day   
1445  696518437233913856    Oh my god 10/10 for every little hot dog pupper   

     expanded_urls  
375            NaN  
707            NaN  
1445           NaN

For the 59 tweets that contain missing data in expanded_urls, 56 are replies or retweets. The remaining 3 tweets (at indexes 375, 707 and 1445) with NaN in the expanded_urls column all have valid ratings but no urls within the text column.

In [21]:
# Checking if these tweets are included in predictions or json_data table

tweet_set = [828361771580813312, 785515384317313025, 696518437233913856]
find_tweet_id = predictions.tweet_id.isin(tweet_set)
sum(find_tweet_id)

0

In [22]:
tweet_set = [828361771580813312, 785515384317313025, 696518437233913856]
find_tweet_id = json_data.id.isin(tweet_set)
json_data[find_tweet_id]

id  favorite_count  retweet_count
374   828361771580813312            2408            195
706   785515384317313025            6800           1477
1443  696518437233913856            4274           1887

These 3 tweets are not in the predictions table, and even though they are included in the json_data table, there is no image url extracted from the json_data.

In [23]:
# Checking rating_numerator and rating_denominator

archive[['rating_numerator', 'rating_denominator']].describe()

rating_numerator  rating_denominator
count       2356.000000         2356.000000
mean          13.126486           10.455433
std           45.876648            6.745237
min            0.000000            0.000000
25%           10.000000           10.000000
50%           11.000000           10.000000
75%           12.000000           10.000000
max         1776.000000          170.000000

For rating_numerator and rating_denominator, the describe() method shows us some quality issues:

The max values are huge: 1776, 170.
The minimum is 0 for both.

In [24]:
archive.rating_denominator.value_counts().sort_index(ascending = False)

170       1
150       1
130       1
120       1
110       1
90        1
80        2
70        1
50        3
40        1
20        2
16        1
15        1
11        3
10     2333
7         1
2         1
0         1
Name: rating_denominator, dtype: int64

In [25]:
sum(archive.rating_denominator != 10)

23

In [26]:
# Checking tweets with denominator =! 10, filtering out replies and retweets

mask_find = (archive.rating_denominator != 10) & \
            archive.in_reply_to_status_id.isna() & \
            archive.retweeted_status_id.isna()

mask_cols = ['text', 'tweet_id', 'rating_numerator', 'rating_denominator']
archive[mask_find][mask_cols]

text            tweet_id  \
433   The floofs have been released I repeat the flo...  820690176645140481   
516   Meet Sam. She smiles 24/7 &amp; secretly aspir...  810984652412424192   
902   Why does this never happen at my front door......  758467244762497024   
1068  After so many requests, this is Bretagne. She ...  740373189193256964   
1120  Say hello to this unbelievably well behaved sq...  731156023742988288   
1165  Happy 4/20 from the squad! 13/10 for all https...  722974582966214656   
1202  This is Bluebert. He just saw that both #Final...  716439118184652801   
1228  Happy Saturday here's 9 puppers on a bench. 99...  713900603437621249   
1254  Here's a brigade of puppers. All look very pre...  710658690886586372   
1274  From left to right:\nCletus, Jerome, Alejandro...  709198395643068416   
1351  Here is a whole flock of puppers.  60/50 I'll ...  704054845121142784   
1433  Happy Wednesday here's a bucket of pups. 44/40...  697463031882764288   
1635  Someone help the girl is being mugged. Several...  684222868335505415   
1662  This is Darrel. He just robbed a 7/11 and is i...  682962037429899265   
1779  IT'S PUPPERGEDDON. Total of 144/120 ...I think...  677716515794329600   
1843  Here we have an entire platoon of puppers. Tot...  675853064436391936   
2335  This is an Albanian 3 1/2 legged  Episcopalian...  666287406224695296   

      rating_numerator  rating_denominator  
433                 84                  70  
516                 24                   7  
902                165                 150  
1068                 9                  11  
1120               204                 170  
1165                 4                  20  
1202                50                  50  
1228                99                  90  
1254                80                  80  
1274                45                  50  
1351                60                  50  
1433                44                  40  
1635               121                 110  
1662                 7                  11  
1779               144                 120  
1843                88                  80  
2335                 1                   2

In [27]:
sum(mask_find)

17

Ignoring the replies and retweets, there are 17 tweets with denominators that aren't 10.

Many of these which are multiples of 10 are about multiple dogs, and the text indicates this, like "floofs", "puppers", "squad" and "pups".

In [28]:
archive.rating_numerator.value_counts().sort_index(ascending = False)

1776      1
960       1
666       1
420       2
204       1
182       1
165       1
144       1
143       1
121       1
99        1
88        1
84        1
80        1
75        2
60        1
50        1
45        1
44        1
27        1
26        1
24        1
20        1
17        1
15        2
14       54
13      351
12      558
11      464
10      461
9       158
8       102
7        55
6        32
5        37
4        17
3        19
2         9
1         9
0         2
Name: rating_numerator, dtype: int64

In [29]:
sum(archive.rating_numerator >= 15)

28

In [30]:
df_tweets_denom_10 = archive[archive.rating_denominator == 10]
df_tweets_denom_10.rating_numerator.value_counts().sort_index(ascending = False)

1776      1
666       1
420       2
182       1
75        2
27        1
26        1
17        1
15        2
14       54
13      351
12      558
11      463
10      461
9       156
8       102
7        54
6        32
5        37
4        15
3        19
2         9
1         8
0         2
Name: rating_numerator, dtype: int64

In [31]:
sum(df_tweets_denom_10.rating_numerator >= 15)

12

There are 12 with numerators >= 15 and denominator = 10. Proceeding to check if choosing 15 as the lower boundary makes sense.

In [32]:
# display tweets with numerator >= 15 that DO have denominator of 10,
# and are NOT replies and are NOT retweets
mask_find = (archive.rating_denominator == 10) & \
            (archive.rating_numerator >= 15) & \
            archive.in_reply_to_status_id.isna() & \
            archive.retweeted_status_id.isna()

mask_cols = ['text', 'tweet_id', 'rating_numerator', 'rating_denominator']
archive[mask_find][mask_cols]

text            tweet_id  \
695   This is Logan, the Chow who lived. He solemnly...  786709082849828864   
763   This is Sophie. She's a Jubilant Bush Pupper. ...  778027034220126208   
979   This is Atticus. He's quite simply America af....  749981277374128128   
1712  Here we have uncovered an entire battalion of ...  680494726643068929   
2074  After so many requests... here you go.\n\nGood...  670842764863651840   

      rating_numerator  rating_denominator  
695                 75                  10  
763                 27                  10  
979               1776                  10  
1712                26                  10  
2074               420                  10

There are 5 tweets (excluding retweets and replies) with denominator = 10 and numerator >= 15. Looking at the text column we see ratings with decimals or huge numbers.

Having only few cases and to avoid imposing a new rating, we will ignore these tweets and drop them.

In [33]:
# Checking tweets with numerator = 0 and denominator = 10, filtering out replies and retweets

mask_find = (archive.rating_denominator == 10) & \
            (archive.rating_numerator == 0) & \
            archive.in_reply_to_status_id.isna() & \
            archive.retweeted_status_id.isna()

mask_cols = ['text', 'tweet_id', 'rating_numerator', 'rating_denominator']
archive[mask_find][mask_cols]

text            tweet_id  \
315  When you're so blinded by your systematic plag...  835152434251116546   

     rating_numerator  rating_denominator  
315                 0                  10

 The tweet with numerator (rating) of '0'.

In [34]:
# Checking 'source'

archive.source.value_counts()

<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     2221
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                          91
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                       33
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>      11
Name: source, dtype: int64

There are 4 types of sources, and they can be simplified by using the display string portion just before the final "<\a>":

* Twitter for iPhone 
* Vine - Make a Scene 
* Twitter Web Client 
* TweetDeck

In [35]:
# Checking 'doggo, floofer, pupper, puppo'

archive.doggo.value_counts()

None     2259
doggo      97
Name: doggo, dtype: int64

In [36]:
archive.floofer.value_counts()

None       2346
floofer      10
Name: floofer, dtype: int64

In [37]:
archive.pupper.value_counts()

None      2099
pupper     257
Name: pupper, dtype: int64

In [38]:
archive.puppo.value_counts()

None     2326
puppo      30
Name: puppo, dtype: int64

In [39]:
archive[archive.puppo == 'puppo'].text

12      Here's a puppo that seems to be on the fence a...
14      This is Stuart. He's sporting his favorite fan...
71      This is Snoopy. He's a proud #PrideMonthPuppo....
94      This is Sebastian. He can't see all the colors...
129     This is Shikha. She just watched you drop a sk...
168     Sorry for the lack of posts today. I came home...
191     Here's a puppo participating in the #ScienceMa...
395     Here's a very loving and accepting puppo. Appe...
397     RT @dog_rates: Say hello to Pablo. He's one go...
398     Say hello to Pablo. He's one gorgeous puppo. A...
413     Here's a super supportive puppo participating ...
438     RT @dog_rates: This is Oliver. He has dreams o...
439     This is Oliver. He has dreams of being a servi...
469     RT @dog_rates: This is Betty. She's assisting ...
546     RT @dog_rates: This is Reginald. He's one magi...
554     This is Diogi. He fell in the pool as soon as ...
567     This is Loki. He'll do your taxes for you. Can...
615     RT @do

## Images Prediction

In [40]:
predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [41]:
predictions.head(5)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [42]:
predictions.p1_dog.value_counts()

True     1532
False     543
Name: p1_dog, dtype: int64

In [43]:
predictions.p2_dog.value_counts()

True     1553
False     522
Name: p2_dog, dtype: int64

In [44]:
predictions.p3_dog.value_counts()

True     1499
False     576
Name: p3_dog, dtype: int64

In [45]:
# check out 20 of the entries that don't predict dogs
predictions[~predictions.p1_dog & \
               ~predictions.p2_dog & \
               ~predictions.p3_dog][['tweet_id', 'p1', 'p1_dog',
                                        'p2', 'p2_dog', 'p3', 'p3_dog']].head(20)

tweet_id                p1  p1_dog                   p2  \
6    666051853826850816        box_turtle   False           mud_turtle   
17   666104133288665088               hen   False                 cock   
18   666268910803644416  desktop_computer   False                 desk   
21   666293911632134144  three-toed_sloth   False                otter   
25   666362758909284353        guinea_pig   False                skunk   
29   666411507551481857              coho   False           barracouta   
45   666786068205871104             snail   False                 slug   
50   666837028449972224       triceratops   False            armadillo   
51   666983947667116034              swab   False            chain_saw   
53   667012601033924608             hyena   False  African_hunting_dog   
56   667065535570550784     jigsaw_puzzle   False              doormat   
69   667188689915760640            vacuum   False                 swab   
73   667369227918143488             teddy   False           bath_towel   
77   667437278097252352         porcupine   False           bath_towel   
78   667443425659232256             goose   False                drake   
93   667549055577362432      electric_fan   False            spotlight   
94   667550882905632768          web_site   False           dishwasher   
96   667724302356258817              ibex   False              bighorn   
98   667766675769573376       fire_engine   False            tow_truck   
100  667782464991965184          lorikeet   False          hummingbird   

     p2_dog                  p3  p3_dog  
6     False            terrapin   False  
17    False           partridge   False  
18    False            bookcase   False  
21    False      great_grey_owl   False  
25    False             hamster   False  
29    False                 gar   False  
45    False               acorn   False  
50    False       common_iguana   False  
51    False                 wig   False  
53    False              coyote   False  
56    False        space_heater   False  
69    False       toilet_tissue   False  
73    False  Christmas_stocking   False  
77    False              badger   False  
78    False                 hen   False  
93    False           lampshade   False  
94    False        oscilloscope   False  
96    False                 ram   False  
98    False                jeep   False  
100   False        African_grey   False

In [46]:
sum(~predictions.p1_dog & \
    ~predictions.p2_dog & \
    ~predictions.p3_dog)

324

Over 500 instances where the algorithm did not predict a dog breed from the image; Combined, there are 324 cases where there is no valid dog breed from any of the three predictions.

## Json_data

In [47]:
json_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
id                2354 non-null int64
favorite_count    2354 non-null int64
retweet_count     2354 non-null int64
dtypes: int64(3)
memory usage: 55.3 KB


In [48]:
json_data.head(5)

id  favorite_count  retweet_count
0  892420643555336193           39467           8853
1  892177421306343426           33819           6514
2  891815181378084864           25461           4328
3  891689557279858688           42908           8964
4  891327558926688256           41048           9774

In [49]:
json_data.describe()

id  favorite_count  retweet_count
count  2.354000e+03     2354.000000    2354.000000
mean   7.426978e+17     8080.968564    3164.797366
std    6.852812e+16    11814.771334    5284.770364
min    6.660209e+17        0.000000       0.000000
25%    6.783975e+17     1415.000000     624.500000
50%    7.194596e+17     3603.500000    1473.500000
75%    7.993058e+17    10122.250000    3652.000000
max    8.924206e+17   132810.000000   79515.000000

## Assessment Summary
### Quality
* Archive
    1. Drop rows with non-null values in 'retweet' data
    2. Drop rows with non-null values in 'replies' data
    3. Drop nulls for 'expanded URLs'
    4. Reformatting 'timestamp' to DateTime
    5. Invalid values in 'name' to "None"
    6. Manually fix of ratings and dropping denominators =! 10
    7. Drop numerators > 15
    8. Establish source category
* Images Prediction
    1. Solve missing values when merging data
### Tidiness
* Archive
    1. Create one column for dog stage and drop invalid category ('floofer')
    2. Drop 'retweet' columns
    3. Drop 'replies' columns
    4. Drop 'rating_denominator' and rename 'rating_numerator' as 'rating'
* Json-data
    1. Rename columns to allow merge smoothly
* Merge json-data and archive
* Images Prediction
    1. Create 'breed' and 'confidence'; merge to archive
    2. Capital letter to 'breed'  
* Order columns

In [50]:
# Creating copies of the dataframes for cleaning
archive_clean = archive.copy()
predictions_clean = predictions.copy()
json_data_clean = json_data.copy()

In [51]:
len_archive = archive_clean.shape[0]; len_archive

2356

# Clean

### Quality
#### Archive #1

#### Define
Drop all rows containing retweets with non-null: retweeted_status_id, retweeted_status_user_id and retweeted_status_timestamp.

#### Code

In [52]:
print("# of retweets:  {}".format(sum(archive_clean.retweeted_status_id.notnull())))

# of retweets:  181


In [53]:
archive_clean = archive_clean[archive_clean.retweeted_status_id.isna()]

#### Test

In [54]:
archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2175 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2175 non-null object
source                        2175 non-null object
text                          2175 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2117 non-null object
rating_numerator              2175 non-null int64
rating_denominator            2175 non-null int64
name                          2175 non-null object
doggo                         2175 non-null object
floofer                       2175 non-null object
pupper                        2175 non-null object
puppo                         2175 non-null object
dtypes: float64(4), int64(3), object(1

In [55]:
# Confirm the rows are dropped; compare with original table length
print("# of rows dropped:  {}".format(len_archive - archive_clean.shape[0]))
# reset the variable for length of the archive
len_archive = archive_clean.shape[0]

# of rows dropped:  181


#### Archive #2

#### Define
Drop all rows that are replies, those that have non-null values in these columns: in_reply_to_status_id and in_reply_to_user_id.

#### Code

In [56]:
print("# of replies:  {}".format(sum(archive_clean.in_reply_to_status_id.notnull())))

# of replies:  78


In [57]:
archive_clean = archive_clean[archive_clean.in_reply_to_status_id.isna()]

#### Test

In [58]:
archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2097 non-null int64
in_reply_to_status_id         0 non-null float64
in_reply_to_user_id           0 non-null float64
timestamp                     2097 non-null object
source                        2097 non-null object
text                          2097 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2094 non-null object
rating_numerator              2097 non-null int64
rating_denominator            2097 non-null int64
name                          2097 non-null object
doggo                         2097 non-null object
floofer                       2097 non-null object
pupper                        2097 non-null object
puppo                         2097 non-null object
dtypes: float64(4), int64(3), object(10)

In [59]:
# Confirm the rows are dropped; compare with previous table length
print("# of rows dropped:  {}".format(len_archive - archive_clean.shape[0]))
# reset the variable for length of the archive
len_archive = archive_clean.shape[0]

# of rows dropped:  78


#### Archive #3

#### Define
Drop tweets with missing data in the expanded_urls.

#### Code

In [60]:
archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2097 non-null int64
in_reply_to_status_id         0 non-null float64
in_reply_to_user_id           0 non-null float64
timestamp                     2097 non-null object
source                        2097 non-null object
text                          2097 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2094 non-null object
rating_numerator              2097 non-null int64
rating_denominator            2097 non-null int64
name                          2097 non-null object
doggo                         2097 non-null object
floofer                       2097 non-null object
pupper                        2097 non-null object
puppo                         2097 non-null object
dtypes: float64(4), int64(3), object(10)

In [61]:
archive_clean = archive_clean[archive_clean.expanded_urls.notnull()]

#### Test

In [62]:
archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2094 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2094 non-null int64
in_reply_to_status_id         0 non-null float64
in_reply_to_user_id           0 non-null float64
timestamp                     2094 non-null object
source                        2094 non-null object
text                          2094 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2094 non-null object
rating_numerator              2094 non-null int64
rating_denominator            2094 non-null int64
name                          2094 non-null object
doggo                         2094 non-null object
floofer                       2094 non-null object
pupper                        2094 non-null object
puppo                         2094 non-null object
dtypes: float64(4), int64(3), object(10)

#### Archive #4

#### Define
Convert timestamp to datetime data type.

#### Code

In [63]:
archive_clean['timestamp'] = pd.to_datetime(archive_clean.timestamp)
archive_clean['timestamp'] = archive_clean.timestamp.dt.floor('s')

#### Test

In [64]:
archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2094 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2094 non-null int64
in_reply_to_status_id         0 non-null float64
in_reply_to_user_id           0 non-null float64
timestamp                     2094 non-null datetime64[ns, UTC]
source                        2094 non-null object
text                          2094 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2094 non-null object
rating_numerator              2094 non-null int64
rating_denominator            2094 non-null int64
name                          2094 non-null object
doggo                         2094 non-null object
floofer                       2094 non-null object
pupper                        2094 non-null object
puppo                         2094 non-null object
dtypes: datetime64[ns, UTC]

#### Archive #5.

#### Define
Replace all lowercase words in the name column with the string "None".

#### Code

In [65]:
names_mask = archive_clean.name.str.contains('^[a-z]', regex = True)
archive_clean[names_mask].name.value_counts().sort_index()

a               55
actually         2
all              1
an               6
by               1
getting          2
his              1
incredibly       1
infuriating      1
just             3
life             1
light            1
mad              1
my               1
not              2
officially       1
old              1
one              4
quite            3
space            1
such             1
the              8
this             1
unacceptable     1
very             4
Name: name, dtype: int64

In [66]:
archive_clean.loc[names_mask, 'name'] = "None"

#### Test

In [67]:
# All these entries were changed to "None"
archive_clean[names_mask].name.value_counts().sort_index()

None    104
Name: name, dtype: int64

#### Archive #6 - Part 1.

#### Define
Manu al fix:

* 740373189193256964: replace 9/11 with 14/10
* 716439118184652801: replace 50/50 with 11/10
* 682962037429899265: replace 7/11 with 10/10
* 666287406224695296: replace 1/2 with 9/10

#### Code

In [68]:
mask_row = archive_clean.tweet_id == 740373189193256964
archive_clean.loc[mask_row, 'rating_numerator'] = 14
archive_clean.loc[mask_row, 'rating_denominator'] = 10

In [69]:
mask_row = archive_clean.tweet_id == 682962037429899265
archive_clean.loc[mask_row, 'rating_numerator'] = 10
archive_clean.loc[mask_row, 'rating_denominator'] = 10

In [70]:
mask_row = archive_clean.tweet_id == 682962037429899265
archive_clean.loc[mask_row, 'rating_numerator'] = 10
archive_clean.loc[mask_row, 'rating_denominator'] = 10

In [71]:
mask_row = archive_clean.tweet_id == 666287406224695296
archive_clean.loc[mask_row, 'rating_numerator'] = 9
archive_clean.loc[mask_row, 'rating_denominator'] = 10

#### Test

In [72]:
mask_cols = ['tweet_id', 'text', 'rating_numerator', 'rating_denominator']
archive_clean[(archive_clean.tweet_id == 740373189193256964) | \
              (archive_clean.tweet_id == 716439118184652801) | \
              (archive_clean.tweet_id == 682962037429899265) | \
              (archive_clean.tweet_id == 666287406224695296) ][mask_cols]

tweet_id                                               text  \
1068  740373189193256964  After so many requests, this is Bretagne. She ...   
1202  716439118184652801  This is Bluebert. He just saw that both #Final...   
1662  682962037429899265  This is Darrel. He just robbed a 7/11 and is i...   
2335  666287406224695296  This is an Albanian 3 1/2 legged  Episcopalian...   

      rating_numerator  rating_denominator  
1068                14                  10  
1202                50                  50  
1662                10                  10  
2335                 9                  10

#### Archive #6 - Part 2.

#### Define
Drop the remaining tweets with rating_denominator values with denominator =! 10.

#### Code

In [73]:
# Checking denominators
archive_clean.rating_denominator.value_counts().sort_index(ascending = False)

170       1
150       1
120       1
110       1
90        1
80        2
70        1
50        3
40        1
20        1
10     2080
7         1
Name: rating_denominator, dtype: int64

In [74]:
archive_clean = archive_clean[archive_clean.rating_denominator == 10]

#### Test

In [75]:
# Checking values, expected only '10'
archive_clean.rating_denominator.value_counts().sort_index(ascending = False)

10    2080
Name: rating_denominator, dtype: int64

#### Archive #7

#### Define
Drop tweets that have rating_numerator > 15.

#### Code

In [76]:
# Checking numerators greater than 15
mask = archive_clean.rating_numerator > 15
archive_clean[mask].rating_numerator.value_counts().sort_index(ascending = False)

1776    1
420     1
75      1
27      1
26      1
Name: rating_numerator, dtype: int64

In [77]:
archive_clean = archive_clean[archive_clean.rating_numerator <= 15]

#### Test

In [78]:
# Checking current numerator values
archive_clean.rating_numerator.value_counts().sort_index(ascending = False)

14     39
13    287
12    485
11    413
10    435
9     153
8      98
7      51
6      32
5      34
4      15
3      19
2       9
1       4
0       1
Name: rating_numerator, dtype: int64

#### Archive #8

#### Define
Replace the source string to get proper source category

#### Code

In [79]:
archive_clean.source.value_counts()

<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     1944
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                          91
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                       30
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>      10
Name: source, dtype: int64

In [80]:
archive_clean['source'] = archive_clean['source'].str.extract('^<a.+>(.+)</a>$')

In [81]:
archive_clean['source'] = archive_clean['source'].astype('category')

#### Test


In [82]:
archive_clean.source.value_counts()

Twitter for iPhone     1944
Vine - Make a Scene      91
Twitter Web Client       30
TweetDeck                10
Name: source, dtype: int64

In [83]:
archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2075 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2075 non-null int64
in_reply_to_status_id         0 non-null float64
in_reply_to_user_id           0 non-null float64
timestamp                     2075 non-null datetime64[ns, UTC]
source                        2075 non-null category
text                          2075 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2075 non-null object
rating_numerator              2075 non-null int64
rating_denominator            2075 non-null int64
name                          2075 non-null object
doggo                         2075 non-null object
floofer                       2075 non-null object
pupper                        2075 non-null object
puppo                         2075 non-null object
dtypes: category(1), date

### Tidiness
#### Archive #1

#### Define
Combine the 3 columns ('doggo', 'pupper', 'puppo') as a new column of dog stages. 
Procediment:

* Create a temporary column called none to store the None values; use lowercase 'n' to avoid confusion with 'None' special NA value.
* Create a new column stage to store the categories: doggo, pupper, puppo, as well as None, and select the stage from the column that contains a value.
* Cast the stage column to categorical data type.
* Drop the 4 original dog stage columns, and the temporary none column.

#### Code

In [84]:
archive_clean[['doggo', 'floofer', 'pupper', 'puppo']].describe()

doggo floofer pupper puppo
count   2075    2075   2075  2075
unique     2       2      2     2
top     None    None   None  None
freq    1992    2065   1847  2051

In [85]:
# Some dogs has multiple stages, will concatenate them.

archive_clean.loc[archive_clean.doggo == 'None', 'doggo'] = ''
archive_clean.loc[archive_clean.floofer == 'None', 'floofer'] = ''
archive_clean.loc[archive_clean.pupper == 'None', 'pupper'] = ''
archive_clean.loc[archive_clean.puppo == 'None', 'puppo'] = ''

archive_clean.groupby(["doggo", "floofer", "pupper", "puppo"]).size().reset_index().rename(columns={0: "count"})

doggo  floofer  pupper  puppo  count
0                                  1741
1                          puppo     23
2                  pupper           219
3         floofer                     9
4  doggo                             72
5  doggo                   puppo      1
6  doggo           pupper             9
7  doggo  floofer                     1

In [86]:
archive_clean['dog_stage'] = archive_clean.doggo + archive_clean.pupper + archive_clean.puppo

# those with more than one category, will be classify as doggo

archive_clean.loc[archive_clean.dog_stage == 'doggopupper', 'dog_stage'] = 'doggo'
archive_clean.loc[archive_clean.dog_stage == 'doggopuppo', 'dog_stage'] = 'doggo'
archive_clean.loc[archive_clean.dog_stage == 'doggofloofer', 'dog_stage'] = 'doggo'

archive_clean.loc[archive_clean.dog_stage == '', 'dog_stage'] = 'None'

In [87]:
# Droping 4 columns
archive_clean = archive_clean.drop(['doggo', 'floofer', 'pupper', 'puppo'], axis = 1)

In [88]:
# Setting the 'stage' column data type to category
archive_clean['dog_stage'] = archive_clean.dog_stage.astype('category')

#### Test

In [89]:
archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2075 entries, 0 to 2355
Data columns (total 14 columns):
tweet_id                      2075 non-null int64
in_reply_to_status_id         0 non-null float64
in_reply_to_user_id           0 non-null float64
timestamp                     2075 non-null datetime64[ns, UTC]
source                        2075 non-null category
text                          2075 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2075 non-null object
rating_numerator              2075 non-null int64
rating_denominator            2075 non-null int64
name                          2075 non-null object
dog_stage                     2075 non-null category
dtypes: category(2), datetime64[ns, UTC](1), float64(4), int64(3), object(4)
memory usage: 215.2+ KB


In [90]:
archive_clean.dog_stage.value_counts()

None      1750
pupper     219
doggo       83
puppo       23
Name: dog_stage, dtype: int64

#### Archive #2
#### Define
Drop all columns related to retweets: retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp.

#### Code

In [91]:
archive_clean = archive_clean.drop(['retweeted_status_id',
                                    'retweeted_status_user_id',
                                    'retweeted_status_timestamp'], axis = 1)

#### Test

In [92]:
# Confirm all retweet related columns are gone
list(archive_clean)

['tweet_id',
 'in_reply_to_status_id',
 'in_reply_to_user_id',
 'timestamp',
 'source',
 'text',
 'expanded_urls',
 'rating_numerator',
 'rating_denominator',
 'name',
 'dog_stage']

#### Archive #3

#### Define
Drop all columns related to replies: in_reply_to_status_id and in_reply_to_user_id.

#### Code

In [93]:
archive_clean = archive_clean.drop(['in_reply_to_status_id',
                                    'in_reply_to_user_id'], axis = 1)

#### Test

In [94]:
# Confirm all reply related columns are gone
list(archive_clean)

['tweet_id',
 'timestamp',
 'source',
 'text',
 'expanded_urls',
 'rating_numerator',
 'rating_denominator',
 'name',
 'dog_stage']

#### Archive #4

#### Define
Drop the 'rating_denominator' column.
Rename the 'rating_numerator' column as 'rating'.
#### Code

In [95]:
archive_clean.rating_denominator.value_counts()

10    2075
Name: rating_denominator, dtype: int64

In [96]:
archive_clean.drop(['rating_denominator'], axis = 1, inplace = True)
archive_clean.rename(index = str, columns = {'rating_numerator': 'rating'}, inplace = True)

#### Test

In [97]:
list(archive_clean)

['tweet_id',
 'timestamp',
 'source',
 'text',
 'expanded_urls',
 'rating',
 'name',
 'dog_stage']

In [98]:
archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2075 entries, 0 to 2355
Data columns (total 8 columns):
tweet_id         2075 non-null int64
timestamp        2075 non-null datetime64[ns, UTC]
source           2075 non-null category
text             2075 non-null object
expanded_urls    2075 non-null object
rating           2075 non-null int64
name             2075 non-null object
dog_stage        2075 non-null category
dtypes: category(2), datetime64[ns, UTC](1), int64(2), object(3)
memory usage: 117.9+ KB


#### Json_data #1

#### Define
Rename Json_data columns.

#### Code

In [99]:
json_data_clean.columns=['tweet_id', 'favourite_count', 'retweet_count']

#### Test

In [100]:
json_data_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
tweet_id           2354 non-null int64
favourite_count    2354 non-null int64
retweet_count      2354 non-null int64
dtypes: int64(3)
memory usage: 55.3 KB


#### Merging Json-data and Archive

#### Define
Merge the retweet_count and favorite_count columns to the archive table, joining on tweet_id.

#### Code 

In [101]:
json_data_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
tweet_id           2354 non-null int64
favourite_count    2354 non-null int64
retweet_count      2354 non-null int64
dtypes: int64(3)
memory usage: 55.3 KB


In [102]:
archive_clean = pd.merge(archive_clean, json_data_clean, 
                         on = 'tweet_id', how = 'left')

#### Test

In [103]:
archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2075 entries, 0 to 2074
Data columns (total 10 columns):
tweet_id           2075 non-null int64
timestamp          2075 non-null datetime64[ns, UTC]
source             2075 non-null category
text               2075 non-null object
expanded_urls      2075 non-null object
rating             2075 non-null int64
name               2075 non-null object
dog_stage          2075 non-null category
favourite_count    2075 non-null int64
retweet_count      2075 non-null int64
dtypes: category(2), datetime64[ns, UTC](1), int64(4), object(3)
memory usage: 150.3+ KB


#### Images predictions #1

#### Define
1. Part 1:

    - Create two new columns in predictions called breed and confidence: check each dog breed prediction flag in order (p1_dog, p2_dog, p3_dog), and copy the breed with the highest confidence level into the breed column.
    - Copy the associated confidence level into the new confidence column.
    - For cases where all 3 prediction flags are False (NOT a valid dog breed), set breed as 'none' and confidence to 0.
2. Part 2:

    - Merge the new breed and confidence columns into the tweet archive table, using an 'inner' join so that we are left with only tweets with images.
    - Now we can fix the confidence column; round the values up to a whole number in percentage format.

#### Code
Part 1

In [104]:
# Setting conditions for selection
conditions = [(predictions_clean['p1_dog'] == True),
              (predictions_clean['p2_dog'] == True),
              (predictions_clean['p3_dog'] == True)]

# Setting choice order based on the selection conditions for predicted breed
choices_breed = [predictions_clean['p1'], 
                 predictions_clean['p2'],
                 predictions_clean['p3']]

# Setting choice order for confidence level based on the selection conditions
choices_confidence = [predictions_clean['p1_conf'], 
                      predictions_clean['p2_conf'], 
                      predictions_clean['p3_conf']]

# Selecting the predicted breed based on the first successful condition
predictions_clean['breed'] = np.select(conditions, choices_breed, 
                                       default = 'none')

# Selecting the predicted confidence level based on the first successful condition
predictions_clean['confidence'] = np.select(conditions, choices_confidence, 
                                            default = 0)

#### Test
Part 1

In [105]:
predictions_clean.head(10)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5  666050758794694657  https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6  666051853826850816  https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7  666055525042405380  https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8  666057090499244032  https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9  666058600524156928  https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   
5        1    Bernese_mountain_dog  0.651137    True    English_springer   
6        1              box_turtle  0.933012   False          mud_turtle   
7        1                    chow  0.692517    True     Tibetan_mastiff   
8        1           shopping_cart  0.962465   False     shopping_basket   
9        1        miniature_poodle  0.201493    True            komondor   

    p2_conf  p2_dog                           p3   p3_conf  p3_dog  \
0  0.156665    True            Shetland_sheepdog  0.061428    True   
1  0.074192    True          Rhodesian_ridgeback  0.072010    True   
2  0.138584    True                   bloodhound  0.116197    True   
3  0.360687    True           miniature_pinscher  0.222752    True   
4  0.243682    True                     Doberman  0.154629    True   
5  0.263788    True   Greater_Swiss_Mountain_dog  0.016199    True   
6  0.045885   False                     terrapin  0.017885   False   
7  0.058279    True                     fur_coat  0.054449   False   
8  0.014594   False             golden_retriever  0.007959    True   
9  0.192305    True  soft-coated_wheaten_terrier  0.082086    True   

                    breed  confidence  
0  Welsh_springer_spaniel    0.465074  
1                 redbone    0.506826  
2         German_shepherd    0.596461  
3     Rhodesian_ridgeback    0.408143  
4      miniature_pinscher    0.560311  
5    Bernese_mountain_dog    0.651137  
6                    none    0.000000  
7                    chow    0.692517  
8        golden_retriever    0.007959  
9        miniature_poodle    0.201493

#### Code
Part 2

In [106]:
archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2075 entries, 0 to 2074
Data columns (total 10 columns):
tweet_id           2075 non-null int64
timestamp          2075 non-null datetime64[ns, UTC]
source             2075 non-null category
text               2075 non-null object
expanded_urls      2075 non-null object
rating             2075 non-null int64
name               2075 non-null object
dog_stage          2075 non-null category
favourite_count    2075 non-null int64
retweet_count      2075 non-null int64
dtypes: category(2), datetime64[ns, UTC](1), int64(4), object(3)
memory usage: 150.3+ KB


In [107]:
# Merging breed and confidence columns to archive
mask_cols = ['tweet_id', 'breed', 'confidence']
archive_clean = pd.merge(archive_clean, predictions_clean[mask_cols],
                         on = 'tweet_id', how = 'inner')

# Changing data type for the confidence interval to a rounded, whole integer (%)
archive_clean.confidence = (archive_clean.confidence * 100).astype(int)

# Droping new columns from the 'predictions' table
predictions_clean.drop(['breed', 'confidence'], axis = 1, inplace = True)

#### Test
Part 2

In [108]:
archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1952 entries, 0 to 1951
Data columns (total 12 columns):
tweet_id           1952 non-null int64
timestamp          1952 non-null datetime64[ns, UTC]
source             1952 non-null category
text               1952 non-null object
expanded_urls      1952 non-null object
rating             1952 non-null int64
name               1952 non-null object
dog_stage          1952 non-null category
favourite_count    1952 non-null int64
retweet_count      1952 non-null int64
breed              1952 non-null object
confidence         1952 non-null int32
dtypes: category(2), datetime64[ns, UTC](1), int32(1), int64(4), object(4)
memory usage: 164.3+ KB


In [109]:
archive_clean.head()

tweet_id                 timestamp              source  \
0  892420643555336193 2017-08-01 16:23:56+00:00  Twitter for iPhone   
1  892177421306343426 2017-08-01 00:17:27+00:00  Twitter for iPhone   
2  891815181378084864 2017-07-31 00:18:03+00:00  Twitter for iPhone   
3  891689557279858688 2017-07-30 15:58:51+00:00  Twitter for iPhone   
4  891327558926688256 2017-07-29 16:00:24+00:00  Twitter for iPhone   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Tilly. She's just checking pup on you....   
2  This is Archie. He is a rare Norwegian Pouncin...   
3  This is Darla. She commenced a snooze mid meal...   
4  This is Franklin. He would like you to stop ca...   

                                       expanded_urls  rating      name  \
0  https://twitter.com/dog_rates/status/892420643...      13   Phineas   
1  https://twitter.com/dog_rates/status/892177421...      13     Tilly   
2  https://twitter.com/dog_rates/status/891815181...      12    Archie   
3  https://twitter.com/dog_rates/status/891689557...      13     Darla   
4  https://twitter.com/dog_rates/status/891327558...      12  Franklin   

  dog_stage  favourite_count  retweet_count               breed  confidence  
0      None            39467           8853                none           0  
1      None            33819           6514           Chihuahua          32  
2      None            25461           4328           Chihuahua          71  
3      None            42908           8964  Labrador_retriever          16  
4      None            41048           9774              basset          55

#### Images predictions #2

#### Define
Chance the names in 'breed' to have the first letter in capital letter

#### Code

In [110]:
archive_clean.breed.unique()

array(['none', 'Chihuahua', 'Labrador_retriever', 'basset',
       'Chesapeake_Bay_retriever', 'Appenzeller', 'Pomeranian',
       'Irish_terrier', 'Pembroke', 'Samoyed', 'French_bulldog',
       'golden_retriever', 'whippet', 'Siberian_husky',
       'Mexican_hairless', 'kuvasz', 'pug', 'Blenheim_spaniel',
       'malamute', 'Italian_greyhound', 'chow', 'German_shepherd',
       'Doberman', 'Eskimo_dog', 'Weimaraner', 'Saluki',
       'miniature_pinscher', 'German_short-haired_pointer',
       'English_springer', 'vizsla', 'bloodhound', 'Bernese_mountain_dog',
       'West_Highland_white_terrier', 'cocker_spaniel',
       'flat-coated_retriever', 'Cardigan', 'Newfoundland',
       'Shetland_sheepdog', 'komondor', 'kelpie',
       'Greater_Swiss_Mountain_dog', 'Border_collie', 'bull_mastiff',
       'Staffordshire_bullterrier', 'Australian_terrier',
       'Bouvier_des_Flandres', 'Bedlington_terrier', 'Boston_bull',
       'black-and-tan_coonhound', 'Great_Dane', 'standard_poodle',
   

In [111]:
archive_clean['breed'] = archive_clean['breed'].str.title()

#### Test 

In [112]:
archive_clean.breed.unique()

array(['None', 'Chihuahua', 'Labrador_Retriever', 'Basset',
       'Chesapeake_Bay_Retriever', 'Appenzeller', 'Pomeranian',
       'Irish_Terrier', 'Pembroke', 'Samoyed', 'French_Bulldog',
       'Golden_Retriever', 'Whippet', 'Siberian_Husky',
       'Mexican_Hairless', 'Kuvasz', 'Pug', 'Blenheim_Spaniel',
       'Malamute', 'Italian_Greyhound', 'Chow', 'German_Shepherd',
       'Doberman', 'Eskimo_Dog', 'Weimaraner', 'Saluki',
       'Miniature_Pinscher', 'German_Short-Haired_Pointer',
       'English_Springer', 'Vizsla', 'Bloodhound', 'Bernese_Mountain_Dog',
       'West_Highland_White_Terrier', 'Cocker_Spaniel',
       'Flat-Coated_Retriever', 'Cardigan', 'Newfoundland',
       'Shetland_Sheepdog', 'Komondor', 'Kelpie',
       'Greater_Swiss_Mountain_Dog', 'Border_Collie', 'Bull_Mastiff',
       'Staffordshire_Bullterrier', 'Australian_Terrier',
       'Bouvier_Des_Flandres', 'Bedlington_Terrier', 'Boston_Bull',
       'Black-And-Tan_Coonhound', 'Great_Dane', 'Standard_Poodle',
   

### Order columns

#### Define
Reorder the column placement: bring numerical columns to the left.
Drop the expanded_urls column.
#### Code

In [113]:
col_list = list(archive_clean.columns)
print(col_list)

['tweet_id', 'timestamp', 'source', 'text', 'expanded_urls', 'rating', 'name', 'dog_stage', 'favourite_count', 'retweet_count', 'breed', 'confidence']


In [114]:
# set the columns in a new order
col_list = ['tweet_id', 'timestamp', 'rating', 'retweet_count', 'favourite_count',
            'dog_stage', 'breed', 'confidence', 'name', 'source', 'text']

archive_clean = archive_clean[col_list]

In [115]:
archive_clean = archive_clean.drop(['text'], axis = 1)

#### Test

In [116]:
archive_clean.sample(5)

tweet_id                 timestamp  rating  retweet_count  \
104   867774946302451713 2017-05-25 16:10:44+00:00      13           7788   
380   811386762094317568 2016-12-21 01:44:13+00:00      11           7444   
1009  706310011488698368 2016-03-06 02:46:44+00:00      12           9034   
1816  668528771708952576 2015-11-22 20:37:34+00:00      12            242   
824   734776360183431168 2016-05-23 16:01:50+00:00      11            608   

      favourite_count dog_stage               breed  confidence    name  \
104             35179      None       Border_Collie          66  Harold   
380             23302    pupper            Pembroke          80   Craig   
1009            23443    pupper            Pembroke          69    None   
1816              496      None  Labrador_Retriever          19  Gòrdón   
824              2742      None      Siberian_Husky          30  Livvie   

                  source  
104   Twitter for iPhone  
380   Twitter for iPhone  
1009  Twitter for iPhone  
1816  Twitter for iPhone  
824   Twitter for iPhone

## Store Data
The data in 'archive_clean' is the result of cleaning the three data tables and merginf them. Thus, the datase is ready to be analysed, henceforce, it will be store into 'twitter_archive_master.csv' file.

In [117]:
archive_clean.to_pickle('twitter_archive_master')